### PR TITLE
Enable Serial Greentea FPGA CI Test Shield tests

### PR DESCRIPTION
**Pull request by fkjagodzinski**
This patch builds on top of #32 and brings:

- `tests/mbed_hal_fpga_ci_test_shield/uart` -- Serial greentea test executed with the FPGA CI Test Shield.
- A static pinmap with fixed `MSTD_CONSTEXPR_FN_14` macro.

### CODE REVIEW

**Comment by hugueskamba** [comment]
Can you please also include this patch?

It helps with ensuring that devices which do not `DEVICE_SERIAL_FC` but do define `DEVICE_SERIAL` compile without error and reduced warnings.
[fpga_shield_uart.patch.txt](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/files/6842749/fpga_shield_uart.patch.txt)

**Comment by fkjagodzinski** [comment]
Fixed the include directive for `DEVICE_SERIAL_FC=0` case and also the unused variable warning.

**Review by ithinuel — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -158,13 +158,13 @@ add_subdirectory(hal)
 # Examples
 add_subdirectory(examples/flash EXCLUDE_FROM_ALL)
 add_subdirectory(examples/gpio EXCLUDE_FROM_ALL)
+add_subdirectory(examples/pwm EXCLUDE_FROM_ALL)
```
[the remark this answers is not in the document]
Comment (fkjagodzinski) [reply CMakeLists.txt]: OK, removed the unrelated clean-up.

**Comment by fkjagodzinski** [comment]
Commits rebased and squashed. This is ready for review and merge. :wink:

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp [hunk 2]
```diff
@@ -209,11 +209,9 @@ static void uart_test_common(int baudrate, int data_bits, SerialParity parity, i
         tx_val = rand() % (1 << test_buff_bits);
         checksum += tx_val;
         serial_putc(&serial, tx_val);
-        us_timestamp_t end_ts = ticker_read_us(us_ticker) + 2 * packet_tx_time;
-        while (tester.rx_get_count() != reps && ticker_read_us(us_ticker) <= end_ts) {
-            // Wait (no longer than twice the time of one packet transfer) for
-            // the FPGA to receive data and update the byte counter.
-        }
+        // Wait twice the time of one packet transfer for
```
Comment (ithinuel) [inline tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:212]: why this change in behaviour when `us_ticker_read` is available?

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp [hunk 3]
```diff
@@ -410,14 +406,16 @@ Case cases[] = {
 utest::v1::status_t greentea_test_setup(const size_t number_of_cases)
 {
     GREENTEA_SETUP(240, "default_auto");
-    srand((unsigned) ticker_read_us(get_us_ticker_data()));
+    us_ticker_init();
+    srand((unsigned) us_ticker_read());
```
Comment (ithinuel) [inline tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:410]: This is deceptive. Init is fairly constant so this seed is not random.

File: tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h [hunk 4]
```diff
@@ -22,6 +22,9 @@
 
 #if DEVICE_SERIAL
```
Comment (ithinuel) [inline tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h:23]: why isn't this occurrence removed as well?

File: tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h [hunk 5]
```diff
@@ -581,9 +581,9 @@ const char *const AnalogoutMaps::name = ANALOGOUT_NAME;
 typedef Port<1, AnalogoutMaps, DefaultFormFactor, TF1> AnalogoutPort;
 #endif
 
-#if DEVICE_SERIAL && 0
-#if DEVICE_SERIAL_FC
+#if DEVICE_SERIAL
```
Comment (ithinuel) [inline tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h:584]: why are they splitted?

File: hal/include/hal/static_pinmap.h [hunk 6]
```diff
@@ -31,8 +31,32 @@
 #include "qspi_api.h"
 #include "ospi_api.h"
 #include "can_api.h"
+
```
Comment (ithinuel) [inline hal/include/hal/static_pinmap.h:34]: can we not add them and remove the include for that c++ header?

File: hal/CMakeLists.txt [hunk 7]
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
         source/mbed_compat.c
         source/mbed_critical_section_api.c
         source/mbed_flash_api.c
-        source/mbed_gpio.c
         source/mbed_gpio_irq.c
+        source/mbed_gpio.c
         # source/mbed_itm_api.c
+        source/mbed_lp_ticker_api.c
```
Comment (ithinuel) [inline hal/CMakeLists.txt:32]: why is this moved? it is not related to this PR.

File: hal/CMakeLists.txt [hunk 8]
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
...
         source/mbed_pinmap_common.c
         source/mbed_pinmap_default.c
         source/mbed_us_ticker_api.c
-        source/mbed_lp_ticker_api.c
-        # source/static_pinmap.cpp
-
         source/mpu/mbed_mpu_v7m.c
         source/mpu/mbed_mpu_v8m.c
+        source/static_pinmap.cpp
```
Comment (ithinuel) [inline hal/CMakeLists.txt:38]: why is this added?

File: hal/CMakeLists.txt [hunk 9]
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
         source/mbed_compat.c
         source/mbed_critical_section_api.c
         source/mbed_flash_api.c
-        source/mbed_gpio.c
         source/mbed_gpio_irq.c
+        source/mbed_gpio.c
```
Comment (ithinuel) [inline hal/CMakeLists.txt:30]: why is it moved?

**Review by ithinuel — CHANGES_REQUESTED** [review]

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp (hunk 2 above)
Comment (fkjagodzinski) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:212]: This is simpler. We only want to check if the char has been received by the test shield in a defined time window. Using `us_ticker_read()` also means ticks->time conversion. I don't like all this extra code here; `wait_us()` is sufficient. I'm also aware this means the test takes longer to complete, but I think we can disregard that.

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp (hunk 3 above)
Comment (fkjagodzinski) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:410]: True; this as bad as what we had before. Anyway, any idea how to improve that?

File: tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h (hunk 4 above)
Comment (fkjagodzinski) [reply tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h:23]: Could you elaborate? I did not remove any `DEVICE_SERIAL`.

File: tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h (hunk 5 above)
Comment (fkjagodzinski) [reply tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h:584]: There are cases where `DEVICE_SERIAL` is defined and `DEVICE_SERIAL_FC` is not. The header has to be there also for these cases.

File: hal/include/hal/static_pinmap.h (hunk 6 above)
Comment (fkjagodzinski) [reply hal/include/hal/static_pinmap.h:34]: I don't follow. Which parts do you want removed?

File: hal/CMakeLists.txt (hunk 9 above)
Comment (fkjagodzinski) [reply hal/CMakeLists.txt:30]: I added an item to the list and then sorted the list. This is must have been in the wrong position and got sorted.

File: hal/CMakeLists.txt (hunk 7 above)
Comment (fkjagodzinski) [reply hal/CMakeLists.txt:32]: This is an effect of sorting the list.

File: hal/CMakeLists.txt (hunk 8 above)
Comment (fkjagodzinski) [reply hal/CMakeLists.txt:38]: The commits are squashed now and I don't see a description for that, I'll check the details. This is probably required by the main.cpp.

Comment (fkjagodzinski) [reply hal/CMakeLists.txt:38]: Good catch! This change should be reverted; the test builds fine without `source/static_pinmap.cpp`. Will fix that.

Comment (fkjagodzinski) [reply hal/CMakeLists.txt]: Fixed via f766fcd.

File: hal/include/hal/static_pinmap.h (hunk 6 above)
Comment (ithinuel) [reply hal/include/hal/static_pinmap.h:34]: line 34 to 59 :)

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp (hunk 2 above)
Comment (ithinuel) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:212]: `wait_us` is deceptive. It hides this code and makes the requirement for the us_ticker.
The verbosity of the previous code isn't satisfying but it is better than `wait_us`.

`wait_us` may not use us_ticker on some platform which make this even more complex and is a pitfall we already encountered in the past weeks.

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp (hunk 3 above)
Comment (ithinuel) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:410]: Not to use random if it is not random.
Define a constant and get rid of occurences of `rand` is the simplest path forward.

File: tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h (hunk 4 above)
Comment (ithinuel) [reply tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h:23]: I meat, with the other recent PR that did remove their related `DEVICE_*`.
If my understanding is correct, in the grand scheme of things, dependencies on peripherals shall be evaluated by CMake and not done in code with `#ifdef`s.

File: hal/include/hal/static_pinmap.h (hunk 6 above)
Comment (fkjagodzinski) [reply hal/include/hal/static_pinmap.h:34]: There are [test cases](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/efab67e6bbe6120d2e0418f16698f21bbd1782dd/tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp#L354) that require [`get_uart_pinmap()`](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/efab67e6bbe6120d2e0418f16698f21bbd1782dd/tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp#L142). This is why we need to `#include "hal/static_pinmap.h"`. This however, declares all it's functions as `MSTD_CONSTEXPR_FN_14` which is not available in our HAL repo. It is provided by the `cxxsupport` in the mbed-os world.

Comment (fkjagodzinski) [reply hal/include/hal/static_pinmap.h:34]: `serial_init_direct()` is still present in our HAL API so should be tested. Adding these defines seems to be the easiest fix for that.

File: tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp (hunk 2 above)
Comment (fkjagodzinski) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:212]: OK, I'll update that not to use `wait_us()`.

Comment (fkjagodzinski) [reply tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp]: Done, 8dc66d008dffd11a79e1d45ba4c74c4d783bcc6c.

**Comment by fkjagodzinski** [comment]
Rebased and squashed the commits. This may be considered ready if we all agree to deal with the unresolved comments in the future. :slightly_smiling_face:

**Review by ithinuel — APPROVED** [review]

**Review by MarceloSalazar — APPROVED** [review]

**Pull request merged**